### PR TITLE
[Enhancement] Optimize mem usage and performance of check strings length

### LINKED ISSUE #11268: [Enhancement] Optimize the mem usage and performance of check strings length
## Enhancement

```
@          0x59b06b2 __gnu_cxx::__verbose_terminate_handler()
    @          0x59af166 __cxxabiv1::__terminate()
    @          0x59af1d1 std::terminate()
    @          0x59af324 __cxa_throw
    @          0x1889744 _Znwm.cold
    @          0x1a0fe51 std::vector<>::reserve()
    @          0x1a0feff starrocks::vectorized::BinaryColumnBase<>::_build_slices()
    @          0x1994ab5 starrocks::vectorized::BinaryColumnBase<>::raw_data()
    @          0x1c785fd starrocks::StringColumnWriter::check_string_lengths()
    @          0x1c7e043 starrocks::StringColumnWriter::append()
    @          0x1b09049 starrocks::SegmentWriter::append_chunk()
    @          0x1cd9e48 starrocks::SegmentRewriter::rewrite()
    @          0x1c4ff0e starrocks::RowsetUpdateState::apply()
    @          0x1a9ca14 starrocks::TabletUpdates::_apply_rowset_commit()
    @          0x1aa0c23 starrocks::TabletUpdates::do_apply()
    @          0x2134b7d starrocks::ThreadPool::dispatch_thread()
    @          0x213038a starrocks::Thread::supervise_thread()
    @     0x7f2d2a3d4dd5 start_thread
    @     0x7f2d299eff6d __clone
```

use 
```
    Slice get_slice(size_t idx) const {
        return Slice(_bytes.data() + _offsets[idx], _offsets[idx + 1] - _offsets[idx]);
    }
```

instead of 

```
template <typename T>
void BinaryColumnBase<T>::_build_slices() const {
    DCHECK(_offsets.size() > 0);
    _slices_cache = false;
    _slices.clear();

    _slices.reserve(_offsets.size() - 1);

    for (size_t i = 0; i < _offsets.size() - 1; ++i) {
        _slices.emplace_back(_bytes.data() + _offsets[i], _offsets[i + 1] - _offsets[i]);
    }

    _slices_cache = true;
}
```

### PR DESCRIPTION
**Pull request by xiaoyong-z**
Signed-off-by: xyz <a997647204@gmail.com>

## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #11268

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

In current implementation, check string length needs to prepare a vector for all slices, which will consumes many memory, also not efficient.

```
    @          0x1a0feff starrocks::vectorized::BinaryColumnBase<>::_build_slices()
    @          0x1994ab5 starrocks::vectorized::BinaryColumnBase<>::raw_data()
    @          0x1c785fd starrocks::StringColumnWriter::check_string_lengths()

template <typename T>
void BinaryColumnBase<T>::_build_slices() const {
    DCHECK(_offsets.size() > 0);
    _slices_cache = false;
    _slices.clear();

    _slices.reserve(_offsets.size() - 1);

    for (size_t i = 0; i < _offsets.size() - 1; ++i) {
        _slices.emplace_back(_bytes.data() + _offsets[i], _offsets[i + 1] - _offsets[i]);
    }

    _slices_cache = true;
}
```

In this pr, we optimize check string length, it doesn't need to create a vector for slices.

### CODE REVIEW

File: be/src/storage/rowset/column_writer.cpp [hunk 1]
```diff
@@ -983,17 +983,18 @@ Status StringColumnWriter::finish() {
...
     const uint8_t* null =
             is_nullable() ? down_cast<const vectorized::NullableColumn*>(&column)->null_column()->raw_data() : nullptr;
-    for (size_t i = 0; i < column.size(); i++) {
+    for (size_t i = 0; i < row_count; i++) {
         // skip string length check if it is null
         if (null != nullptr && null[i] == starrocks::vectorized::DATUM_NULL) {
             continue;
         }
-        if (slices[i].get_size() > limit) {
+        Slice slice = column.get_slice(i);
```
Comment (imay) [inline be/src/storage/rowset/column_writer.cpp:994]: Better to leave a comment to avoid someone change it back.

**Review by trueeyu — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by xiaoyong-z** [comment]
@mergify backport branch-2.4

**Comment by xiaoyong-z** [comment]
@mergify backport branch-2.3

**Comment by trueeyu** [comment]
@mergify backport branch-2.2

**Comment by trueeyu** [comment]
@mergify backport branch-2.2-yunlu

**Pull request merged**